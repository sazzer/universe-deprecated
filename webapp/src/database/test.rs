use lazy_static::lazy_static;
use log::info;
use postgres::{Client, NoTls};
use testcontainers::{clients::Cli, images::postgres::Postgres, Container, Docker};

lazy_static! {
    static ref DOCKER: Cli = { Cli::default() };
}

pub struct TestDatabase<'d> {
    #[allow(dead_code)]
    node: Container<'d, Cli, Postgres>,
    url: String,
}

impl<'d> TestDatabase<'d> {
    pub fn new() -> Self {
        let node = DOCKER.run(Postgres::default());

        let host_port = node.get_host_port(5432).unwrap();
        let url = format!("postgres://postgres:postgres@localhost:{}", host_port);
        info!("Running postgres on {}", url);

        TestDatabase { node, url }
    }

    pub fn client(&self) -> Client {
        Client::connect(&self.url, NoTls).unwrap()
    }
}
