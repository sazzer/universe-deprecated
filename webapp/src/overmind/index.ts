import { namespaced } from "overmind/config";
import { createHook } from "overmind-react";
import { IConfig } from "overmind";
import * as login from "../login";
import * as authentication from "../authentication";

export const config = namespaced({
  login,
  authentication
});

export const useOvermind = createHook<typeof config>();

declare module "overmind" {
  // tslint:disable:interface-name
  interface Config extends IConfig<typeof config> {}
}
