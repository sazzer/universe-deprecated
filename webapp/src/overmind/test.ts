import { createOvermindMock, OvermindMock, Config } from "overmind";
import { config } from "../overmind";
import { cloneDeep, merge } from "lodash";

/**
 * NestedPartial is stolen from Overmind to represent a part of the Overmind structure
 */
export declare type NestedPartial<T> = T extends Function
  ? T
  : Partial<
      {
        [P in keyof T]: NestedPartial<T[P]>;
      }
    >;

/**
 * Create a test instance of the Overmind to work with
 * @param mockEffects The mock effects to wire in
 * @param initialState The initial state to wire in
 */
export function createTestOvermind(
  mockEffects?: NestedPartial<Config["effects"]>,
  initialState?: NestedPartial<Config["state"]>
): OvermindMock<Config> {
  const mergedState = merge(cloneDeep(config), { state: initialState });
  return createOvermindMock(mergedState, mockEffects);
}
