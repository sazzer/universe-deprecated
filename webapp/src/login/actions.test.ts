import { createOvermindMock, OvermindMock, Config } from "overmind";
import { config } from "../overmind";
import { cloneDeep, merge } from "lodash";
import { createTestOvermind } from "../overmind/test";

describe("checkUsername", () => {
  let checkUsernameEffect: jest.Mock;
  let overmind: OvermindMock<Config>;

  beforeEach(() => {
    checkUsernameEffect = jest.fn();

    overmind = createTestOvermind({
      login: {
        api: {
          checkUsername: checkUsernameEffect
        }
      }
    });
  });
  test("When the username is known", async () => {
    checkUsernameEffect.mockReturnValueOnce(Promise.resolve(true));
    await overmind.actions.login.checkUsername("testuser");
    expect(overmind.mutations).toMatchSnapshot();
    expect(overmind.state).toMatchSnapshot();
  });
  test("When the username is not known", async () => {
    checkUsernameEffect.mockReturnValueOnce(Promise.resolve(false));
    await overmind.actions.login.checkUsername("testuser");
    expect(overmind.mutations).toMatchSnapshot();
    expect(overmind.state).toMatchSnapshot();
  });
  test("When the server call fails", async () => {
    checkUsernameEffect.mockReturnValueOnce(
      Promise.reject(new Error("Network Error"))
    );
    await overmind.actions.login.checkUsername("testuser");
    expect(overmind.mutations).toMatchSnapshot();
    expect(overmind.state).toMatchSnapshot();
  });
});

describe("resetLogin", () => {
  test("Resetting from the initial state", () => {
    const overmind = createTestOvermind();
    overmind.actions.login.resetLogin();
    expect(overmind.mutations).toMatchSnapshot();
    expect(overmind.state).toMatchSnapshot();
  });
  test("Resetting from the registering state", () => {
    const overmind = createTestOvermind(
      {},
      {
        login: {
          mode: {
            current: "registering"
          }
        }
      }
    );

    overmind.actions.login.resetLogin();
    expect(overmind.mutations).toMatchSnapshot();
    expect(overmind.state).toMatchSnapshot();
  });
  test("Resetting from the authenticating state", () => {
    const overmind = createTestOvermind(
      {},
      {
        login: {
          mode: {
            current: "authenticating"
          }
        }
      }
    );
    overmind.actions.login.resetLogin();
    expect(overmind.mutations).toMatchSnapshot();
    expect(overmind.state).toMatchSnapshot();
  });
});

describe("cancelLogin", () => {
  test("Cancelling from the initial state", () => {
    const overmind = createTestOvermind();
    overmind.actions.login.cancelLogin();
    expect(overmind.mutations).toMatchSnapshot();
    expect(overmind.state).toMatchSnapshot();
  });
  test("Cancelling from the registering state", () => {
    const overmind = createTestOvermind(
      {},
      {
        login: {
          mode: {
            current: "registering"
          }
        }
      }
    );
    overmind.actions.login.cancelLogin();
    expect(overmind.mutations).toMatchSnapshot();
    expect(overmind.state).toMatchSnapshot();
  });
  test("Cancelling from the authenticating state", () => {
    const overmind = createTestOvermind(
      {},
      {
        login: {
          mode: {
            current: "authenticating"
          }
        }
      }
    );
    overmind.actions.login.cancelLogin();
    expect(overmind.mutations).toMatchSnapshot();
    expect(overmind.state).toMatchSnapshot();
  });
});
