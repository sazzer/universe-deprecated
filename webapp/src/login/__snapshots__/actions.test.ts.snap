// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cancelLogin Cancelling from the authenticating state 1`] = `
Array [
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.error",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.error",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "initial",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "initial",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
]
`;

exports[`cancelLogin Cancelling from the authenticating state 2`] = `
Object {
  "login": Object {
    "error": null,
    "isAuthenticating": false,
    "isLoading": false,
    "isRegistering": false,
    "mode": StateMachine {
      "authenticating": [Function],
      "current": "initial",
      "initial": [Function],
      "loading": [Function],
      "registering": [Function],
      Symbol(CURRENT_EXIT): undefined,
    },
    "username": null,
  },
}
`;

exports[`cancelLogin Cancelling from the initial state 1`] = `
Array [
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.error",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.error",
    "revert": [Function],
  },
]
`;

exports[`cancelLogin Cancelling from the initial state 2`] = `
Object {
  "login": Object {
    "error": null,
    "isAuthenticating": false,
    "isLoading": false,
    "isRegistering": false,
    "mode": StateMachine {
      "authenticating": [Function],
      "current": "initial",
      "initial": [Function],
      "loading": [Function],
      "registering": [Function],
    },
    "username": null,
  },
}
`;

exports[`cancelLogin Cancelling from the registering state 1`] = `
Array [
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.error",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.error",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "initial",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "initial",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
]
`;

exports[`cancelLogin Cancelling from the registering state 2`] = `
Object {
  "login": Object {
    "error": null,
    "isAuthenticating": false,
    "isLoading": false,
    "isRegistering": false,
    "mode": StateMachine {
      "authenticating": [Function],
      "current": "initial",
      "initial": [Function],
      "loading": [Function],
      "registering": [Function],
      Symbol(CURRENT_EXIT): undefined,
    },
    "username": null,
  },
}
`;

exports[`checkUsername When the server call fails 1`] = `
Array [
  Object {
    "args": Array [
      "loading",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.error",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "initial",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "initial",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "Error: Network Error",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.error",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "Error: Network Error",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.error",
    "revert": [Function],
  },
]
`;

exports[`checkUsername When the server call fails 2`] = `
Object {
  "login": Object {
    "error": "Error: Network Error",
    "isAuthenticating": false,
    "isLoading": false,
    "isRegistering": false,
    "mode": StateMachine {
      "authenticating": [Function],
      "current": "initial",
      "initial": [Function],
      "loading": [Function],
      "registering": [Function],
      Symbol(CURRENT_EXIT): undefined,
    },
    "username": null,
  },
}
`;

exports[`checkUsername When the username is known 1`] = `
Array [
  Object {
    "args": Array [
      "loading",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.error",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "authenticating",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "authenticating",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "testuser",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "testuser",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
]
`;

exports[`checkUsername When the username is known 2`] = `
Object {
  "login": Object {
    "error": null,
    "isAuthenticating": false,
    "isLoading": false,
    "isRegistering": false,
    "mode": StateMachine {
      "authenticating": [Function],
      "current": "authenticating",
      "initial": [Function],
      "loading": [Function],
      "registering": [Function],
      Symbol(CURRENT_EXIT): undefined,
    },
    "username": "testuser",
  },
}
`;

exports[`checkUsername When the username is not known 1`] = `
Array [
  Object {
    "args": Array [
      "loading",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.error",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "registering",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "registering",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "testuser",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "testuser",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
]
`;

exports[`checkUsername When the username is not known 2`] = `
Object {
  "login": Object {
    "error": null,
    "isAuthenticating": false,
    "isLoading": false,
    "isRegistering": false,
    "mode": StateMachine {
      "authenticating": [Function],
      "current": "registering",
      "initial": [Function],
      "loading": [Function],
      "registering": [Function],
      Symbol(CURRENT_EXIT): undefined,
    },
    "username": "testuser",
  },
}
`;

exports[`resetLogin Resetting from the authenticating state 1`] = `
Array [
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.error",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "initial",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
]
`;

exports[`resetLogin Resetting from the authenticating state 2`] = `
Object {
  "login": Object {
    "error": null,
    "isAuthenticating": false,
    "isLoading": false,
    "isRegistering": false,
    "mode": StateMachine {
      "authenticating": [Function],
      "current": "initial",
      "initial": [Function],
      "loading": [Function],
      "registering": [Function],
      Symbol(CURRENT_EXIT): undefined,
    },
    "username": null,
  },
}
`;

exports[`resetLogin Resetting from the initial state 1`] = `
Array [
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.error",
    "revert": [Function],
  },
]
`;

exports[`resetLogin Resetting from the initial state 2`] = `
Object {
  "login": Object {
    "error": null,
    "isAuthenticating": false,
    "isLoading": false,
    "isRegistering": false,
    "mode": StateMachine {
      "authenticating": [Function],
      "current": "initial",
      "initial": [Function],
      "loading": [Function],
      "registering": [Function],
    },
    "username": null,
  },
}
`;

exports[`resetLogin Resetting from the registering state 1`] = `
Array [
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.error",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "initial",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
]
`;

exports[`resetLogin Resetting from the registering state 2`] = `
Object {
  "login": Object {
    "error": null,
    "isAuthenticating": false,
    "isLoading": false,
    "isRegistering": false,
    "mode": StateMachine {
      "authenticating": [Function],
      "current": "initial",
      "initial": [Function],
      "loading": [Function],
      "registering": [Function],
      Symbol(CURRENT_EXIT): undefined,
    },
    "username": null,
  },
}
`;
